File: solution/108/solution.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        if len(nums) == 0:return None
        nodelist = []
        self.preorder_insert(nums, nodelist, 0, len(nums)-1)
        for i in xrange(1, len(nodelist)):
            self.BSTinsert(nodelist[0], nodelist[i])
        return nodelist[0]

    def preorder_insert(self, nums, nodelist, start, end):
        if start > end:
            return
        mid = (start + end)/2
        nodelist.append(TreeNode(nums[mid]))
        self.preorder_insert(nums, nodelist, start, mid-1)
        self.preorder_insert(nums, nodelist, mid+1, end)

    def BSTinsert(self, node, insertnode):
        if insertnode.val < node.val:
            if node.left:
                self.BSTinsert(node.left, insertnode)
            else:
                node.left = insertnode
        else:
            if node.right:
                self.BSTinsert(node.right, insertnode)
            else:
                node.right = insertnode

s = Solution()
print s.sortedArrayToBST([[-10,-3,0,5,9]])

File: solution/108/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void BSTinsert(TreeNode *node, TreeNode *insert_node){
        if(insert_node->val < node->val){
            if(node->left){
                BSTinsert(node->left, insert_node);
            }else{
                node->left = insert_node;
            }
        }else{
            if(node->right){
                BSTinsert(node->right, insert_node);
            }else{
                node->right = insert_node;
            }
        }
    }

    void preorder_insert(std::vector<int> &nums, std::vector<TreeNode*> &nodelist, int begin, int end){
        if(end<begin)return;
        int mid = (begin + end) / 2;
        nodelist.push_back(new TreeNode(nums[mid]));
        preorder_insert(nums, nodelist, begin, mid-1);
        preorder_insert(nums, nodelist, mid+1, end);
    }

    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if(nums.size()==0)return NULL;
        std::vector<TreeNode *> nodelist;
        preorder_insert(nums, nodelist, 0, nums.size() - 1);
        for(int i=1;i<nodelist.size();i++){
            BSTinsert(nodelist[0], nodelist[i]);
        }
        return nodelist[0];
    }


};
